### PR TITLE
Delegation update: incentives

### PR DESCRIPTION
**Pull request by kantp**

### CODE REVIEW

**Comment by mgudemann** [comment]
@kantp there has been a repo reorganization the doc now lives in `shelley/design-spec/`

File: shelley/design-spec/delegation_design_spec.tex [hunk 1]
```diff
@@ -2523,68 +2527,77 @@ \subsection{Overview of Incentives}
...
-  to operate several pools. It will also help preventing Sybil attacks,
-  where an attacker with low stake tries to gain control over a majority
-  of stake by creating a lot of pools with low costs.)
+  Pool rewards will slightly increase with the stake their operator contributes to
+  the pool. There is no minimal stake required to create a pool -- anybody can
+  do this. However, pools led by operators with high stake will get higher
+  rewards. (This will discourage pool operators from splitting their stake to
+  operate several pools. It will also help preventing Sybil attacks, where an
+  attacker with low stake tries to gain control over a majority of stake by
+  creating a lot of pools with low costs.)
```
Comment (vantuz-subhuman) [inline shelley/design-spec/delegation_design_spec.tex]: Shouldn't it say "owner" whenever pledge/stake is mentioned? The rest of the spec seems to emphasising how pool operator is different from people funding the pool.

File: shelley/design-spec/delegation_design_spec.tex [hunk 2]
```diff
@@ -2652,23 +2665,47 @@ \subsubsection{Transaction Fees}
...
+\item[T] is the amount of ada in circulation at the beginning of the epoch for
+  which we want to calculate the rewards pot.
+\end{itemize}
+
+The dependece on \(\eta\) incentivises cooperative behaviour, and in particular
+discourages the pools sabotaging each others blocks. Note that \(\eta\) can
+exceed \(1\) when there are more blocks produced in an epoch than would be
+expected on average. But the product \(\min(\eta, 1)\rho\) is always bounded by
+\(1\), which is necessary to ensure we never exceed \(T_\infty\) ada in
+circulation.
```
Comment (vantuz-subhuman) [inline shelley/design-spec/delegation_design_spec.tex]: Beautiful! ❤️

File: shelley/design-spec/delegation_design_spec.tex [hunk 3]
```diff
@@ -2652,23 +2665,47 @@ \subsubsection{Transaction Fees}
...
-future. At this moment, \(T=31,000,000,000\) and
-\(T_\infty=45,000,000,000\). Then the amount of
-\(\rho\cdot(T_\infty - T)\) ADA will be newly created.
+Every epoch, the total amount of ada in circulation \(T\) will be increased by
+adding ada to the rewards pot. To ensure the amount of ada never exceeds a
+finite, specified limit \(T_\infty\), the increase will reduce exponentially.
+Also, the increase will depend on the number of blocks that have been produced
+during the epoch. Specifically, each epoch, the contribution from monetary
+expansion to the rewards pot is given by
+\[\min(\eta, 1) \rho \left(T_\infty - T\right)\,,\]
```
Comment (vantuz-subhuman) [inline shelley/design-spec/delegation_design_spec.tex]: Just to clarify. What is the purpose of `min(η,1)`? Is it possible that created number of block is *larger* than expected number of blocks in an epoch? Is it because of slots with no leaders in Praos? Do I get it right that nodes will know the expected approximate number of leader-less slots?

Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: Exactly! In Praos, you can have more blocks in an epoch than you would expect on average.

File: shelley/design-spec/delegation_design_spec.tex (hunk 1 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: Right, thanks!

Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: Added a commit to consistently use "owner's stake".

Comment (vantuz-subhuman) [reply shelley/design-spec/delegation_design_spec.tex]: Thank you!

File: shelley/design-spec/delegation_design_spec.tex [hunk 4]
```diff
@@ -2752,110 +2864,66 @@ \subsubsection{Pool Rewards}
...
+\item
+  We will always have
+  \[ \sum_{\text{pools}} \hat{f}(s,\sigma,\pbar) \leq R \,, \]
+  so that all actual rewards can be paid from the rewards pot\footnote{ To prove
+    this, use that \(s' \leq \sigma' \leq z_0\), and
+    \(\sum_{\text{pools}}\beta=1\).}.
+
+  The difference \(R - \sum_{\text{pools}} \hat{f}(s,\sigma,\pbar)\) will be
+  sent to the treasury.
+\end{itemize}
```
Comment (vantuz-subhuman) [inline shelley/design-spec/delegation_design_spec.tex]: image

The `∑f(s,σ,p)` represents all reward calculations, including capping pool share by `1/k`, right? Does this mean that different between the idea pool reward and capped pool reward (for over-saturated pools) is also sent to treasury?

Because I think I remember discussions that difference caused by over-saturation is carried-over to the next epoch, and difference cause by performance is sent to the treasury. But now it also seems to be missing in the spec itself 🤔 

Spec says this:

image

So it means:
1. We make total pool
2. We subtract what goes to the treasury
3. We subtract the capped pool rewards
4. The rest also goes to treasury

It sounds like there's no carry-over to the next epoch anymore in the system. If so, then what does this arrow back from total pool to rewards represent?

image

Where does carried-over value come from, if difference between `R` and total pool rewards is always fully sent to the treasury?

Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: Thanks for carefully reviewing this!

The reason for the change is this: in the original scheme, we had the hierarchy that

`rewards pot >= sum of maximal rewards >= sum of actual rewards`

The first inequality is rather straightforward to prove, and the second was trivial, since each reward was multiplied by a factor `<=1`.

But now, the apparent performance can be `>1`, and there is no reason for the second inequality to hold anymore. It's not hard to come up with counterexamples: just take one pool that is oversaturated and is chosen to produce fewer blocks than expected on average, and other pools that are not saturated are chosen to produce more. In this case, the sum of the actual rewards will be larger than the sum of the optimal rewards (note that we're not calling them maximal rewards anymore).

Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: Ah, I think this means there is no contribution to the rewards pot anymore from the last epoch, so I should better go and correct this in 6.4.

File: shelley/design-spec/delegation_design_spec.tex [hunk 5]
```diff
@@ -2499,19 +2503,19 @@ \section{Design of Incentives}
 \subsection{Overview of Incentives}
 \label{overview-of-incentives}
 
-On a high level, goal of the incentives mechanism is to incentivise
-stakeholders to follow the protocol and thereby guaranteeing secure and
+On a high level, the goal of the incentives mechanism is to incentivise
+stakeholders to follow the protocol and thereby to guarantee the secure and
```
Comment (dnadales) [inline shelley/design-spec/delegation_design_spec.tex]: FWIW I liked the gerund better, but I'm not a native speaker :)

File: shelley/design-spec/delegation_design_spec.tex [hunk 6]
```diff
@@ -2596,16 +2611,16 @@ \subsection{Parameters}
...
+    influence.
 \item
   The \emph{expansion rate} \(\rho\in[0,1]\), determining the fraction
-  of still available ADA that will be created per epoch.
+  of still available ada that will be created per epoch.
 \item
   The fraction \(\tau\in[0,1]\) of rewards going to the treasury.
-\item
-  An exponent \(\gamma\in(0,\infty)\) for penalty calculation.
+%% \item
```
Comment (dnadales) [inline shelley/design-spec/delegation_design_spec.tex]: Any reason why we're leaving this comment here?

File: shelley/design-spec/delegation_design_spec.tex [hunk 7]
```diff
@@ -2644,31 +2659,55 @@ \subsection{Reminder: Stakepool Registration}
...
 \label{monetary-expansion}
 
-Let \(T\) be the total amount of ADA in existence during a specific
-epoch, and let \(T_\infty\) be the maximal possible amount of ADA in the
-future. At this moment, \(T=31,000,000,000\) and
-\(T_\infty=45,000,000,000\). Then the amount of
-\(\rho\cdot(T_\infty - T)\) ADA will be newly created.
+Every epoch, the total amount of ada in circulation \(T\) will be increased by
+adding ada to the rewards pot. To ensure the amount of ada never exceeds a
+finite, specified limit \(T_\infty\), the increase will reduce exponentially.
```
Comment (dnadales) [inline shelley/design-spec/delegation_design_spec.tex]: It doesn't belong to this PR probably, but I find it a bit confusing that we use the infinity symbol for a finite value :thinking:

File: shelley/design-spec/delegation_design_spec.tex [hunk 8]
```diff
@@ -2718,32 +2747,108 @@ \subsection{Reward Splitting}
...
+\[\beta = \sigma p r_e \,,\]
+where
+\begin{enumerate}
+\item[\(\sigma\)] is the relative stake \(\sigma\) of the pool (the fraction of
+  the absolute stake that the pool controls),
+\item[\(p\)] is the performance \(p\) of the pool, i.e. the fraction
+  \[
+  p = \frac{n}{\max(N,1)}
+  \]
+  of the number \(n\) blocks it successfully added to the chain and the number
```
Comment (dnadales) [inline shelley/design-spec/delegation_design_spec.tex]: I don't get this, is `N < n`? Is `n` the number of blocks successfully added to the chain by the pool? or by all the pools?

File: shelley/design-spec/delegation_design_spec.tex [hunk 9]
```diff
@@ -2718,32 +2747,108 @@ \subsection{Reward Splitting}
...
+\[
+\beta = \sigma \frac{n}{\max(N,1)} \frac{N}{\sigma \overline{N}}
+= \frac{n}{\overline{N}}\,,
+\]
+which is indeed the fraction of blocks produced by the pool\footnote{When
+  cancelling \(N\) and \(\max(N,1)\), note that for \(N=0\), also \(n=0\), and
+  the equation trivially holds.}.
+
+In Ouroboros Praos, we can observe \(\sigma\), \(n\), and \(\overline{N}\), but
+we do not have a direct handle on any of \(N\), \(r_e\), or \(p\). From the
```
Comment (dnadales) [inline shelley/design-spec/delegation_design_spec.tex]: Not "having a direct handle on ..." means "we cannot observe ..."?

File: shelley/design-spec/delegation_design_spec.tex [hunk 10]
```diff
@@ -2718,32 +2747,108 @@ \subsection{Reward Splitting}
...
+observables, we can extract
+\[ \beta = \frac{n}{\overline{N}}\,. \]
+While the true performance \(p\) is not accessible, we can define and measure
+the \emph{apparent performance}
```
Comment (dnadales) [inline shelley/design-spec/delegation_design_spec.tex]: To me it is not clear (from the formulae above) why this performance is "apparent". I'd expect that "apparent" in this context means "approximation" (but maybe I got this wrong). If that's the case, then I wonder where are we loosing precision, since we replaced equals by equals in the derivation of `\pbar`.

File: shelley/design-spec/delegation_design_spec.tex [hunk 11]
```diff
@@ -2965,31 +3025,75 @@ \subsubsection{Non-Myopic Pool Stake}
...
+insert the apparent performance averaged over several epochs as \(\pbar\), not
+the value of \(\pbar\) from just the last epoch.
+
+Averaging can be done efficiently via an exponential moving average, or by using
+a moving window and keeping track of the apparent performance during the last
+couple of epochs. Note that since we perform averaging only when predicting the
+rewards to allow stakeholders to make an informed decision, but not when
+actually dealing out the rewards, it will be fine if different wallets will
+choose different methods of calculating the averages, since it does not affect
+the consensus.
```
Comment (dnadales) [inline shelley/design-spec/delegation_design_spec.tex]: Nice :)

File: shelley/design-spec/delegation_design_spec.tex (hunk 4 above)
Comment (vantuz-subhuman) [reply shelley/design-spec/delegation_design_spec.tex]: Ahhhhh, so pools are over-rewarded for over-performance? Yeah, I think it was not present in the previous model.

So just to clarify - if we have `K=2` and two pools: `A` with 75% stake (so over-saturated above 50%) and `B` with 25% stake (under-saturated), and we have 100 slots to assign, and each pool gets assigned 50 slots.

`A` is expected to produce 75 blocks, but produces 50, so it received `(0.5*R)*0.66` = `~0.33*R`

`B` is expected to produce 25 blocks, but produces 50, so it receives `(0.25*R)*2` = `0.5*R`

Total distributed = `0.83*R`
Leftover `0.17*R` is sent to treasury?

My question assumed that older version of the spec said something like:
1. Ideal reward for `A` would be `0.75*R` (calculated from stake share)
2. Capped reward for `A` is `(0.5*R)` (because of `K=2`)
3. Diff is `0.25*R` is sent to next epoch
4. Slashed for performance reward for `A` is `(0.66*0.5*R)` = `0.33*R`
5. Diff is `0.17*R` is sent to treasury

And for `B` calculation would imply that it *never* can get reward larger than `0.25*R`, because of its stake. But now it is implied that `B` will get more than its share of rewards because of over-performing? Am I getting it right?

Comment (vantuz-subhuman) [reply shelley/design-spec/delegation_design_spec.tex]: Ok, I get this case, so if some pool is over-performing - it gets piece of rewards that would otherwise be sent to next-epoch or to treasury. But my question is about the basic case: if we have `K=2`, 100 slots, and two pools: `A` at 75% stake (over) and `B` at 25% stake (under). Randomness is fair in this epoch and they get assigned expected number of slots. So:

`A` produces 75 slots, but gets `(0.5*R)` rewards, because of `K=2`

`B` produces 25 slots and gets `(0.25*R)` rewards.

There's `(0.25*R)` left-over from capping `A`. Does this gets sent to treasury in this case? Why?

Because I seem to remember it was planned to carry-over this value to next epoch.

File: shelley/design-spec/delegation_design_spec.tex (hunk 6 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: Nope, no reason. I'll delete it :)

File: shelley/design-spec/delegation_design_spec.tex (hunk 7 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: It's the amount of ada in the limit time -> infinity.

File: shelley/design-spec/delegation_design_spec.tex (hunk 9 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: Yes. Should I add a clarification?

File: shelley/design-spec/delegation_design_spec.tex (hunk 8 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: No, we have `n <= N`. `n` is the number of blocks this pool produced, `N` is the number of blocks it _should_ have produced. So the performance is bounded by `1`.

File: shelley/design-spec/delegation_design_spec.tex (hunk 10 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: It's the product of the actual performance of the pool and all the random influences on the number of blocks that the pool produces -- their luck or bad luck in the leader election, network effects, ...

The reason I called it apparent performance is that if you are oblivious to the randomness in leader election and assume a perfect network, the pool would appear to have this performance.

Does that make things clearer :)

File: shelley/design-spec/delegation_design_spec.tex (hunk 11 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: Isn't it :)

File: shelley/design-spec/delegation_design_spec.tex (hunk 4 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: The point is that generally, we will no longer have

`sum of actual rewards <= sum of optimal reward`,

so if we were to send

`reward pot - sum of optimal rewards`

to the next epoch's reward pot, we might not have enough funds to pay out all rewards.

We could make a case distinction, and send

`min (sum of optimal rewards, sum of actual rewards)`

over to the next epoch, and

`max(0, (sum of optimal rewards) - (sum of actual rewards))`

to the treasury, I guess. But I don't see a compelling reason why we should do that. It would increase the rewards for everyone in the presence of oversaturated pools. We don't really anticipate having oversaturated pools, and if we do see them, we want people to act to get away from them.

Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: And the only way to ensure that the actual rewards are bounded by the optimal rewards would be capping the factor from the apparent performance at 1. But that would mean that we'll systematically underpay everyone.

Comment (vantuz-subhuman) [reply shelley/design-spec/delegation_design_spec.tex]: Understood. Thank you very much, @kantp! 🙌

File: shelley/design-spec/delegation_design_spec.tex [hunk 12]
```diff
@@ -2644,31 +2659,55 @@ \subsection{Reminder: Stakepool Registration}
...
+\item[T] is the amount of ada in circulation at the beginning of the epoch for
+  which we want to calculate the rewards pot.
+\end{itemize}
+
+The dependece on \(\eta\) incentivises cooperative behaviour, and in particular
+discourages the pools sabotaging each others blocks. Note that \(\eta\) can
+exceed \(1\) when there are more blocks produced in an epoch than would be
+expected on average. But the product \(\min(\eta, 1)\rho\) is always bounded by
+\(1\), which is necessary to ensure we never exceed \(T_\infty\) ada in
+circulation.
```
Comment (brunjlar) [inline shelley/design-spec/delegation_design_spec.tex]: I think we should mention how \eta is actually calculated (provided it is _not_ simply the number of slots per epoch).

File: shelley/design-spec/delegation_design_spec.tex [hunk 13]
```diff
@@ -2644,31 +2659,55 @@ \subsection{Reminder: Stakepool Registration}
...
-future. At this moment, \(T=31,000,000,000\) and
-\(T_\infty=45,000,000,000\). Then the amount of
-\(\rho\cdot(T_\infty - T)\) ADA will be newly created.
+Every epoch, the total amount of ada in circulation \(T\) will be increased by
+adding ada to the rewards pot. To ensure the amount of ada never exceeds a
+finite, specified limit \(T_\infty\), the increase will reduce exponentially.
+Also, the increase will depend on the number of blocks that have been produced
+during the epoch. Specifically, each epoch, the contribution from monetary
+expansion to the rewards pot is given by
+\[\min(\eta, 1) \rho \left(T_\infty - T\right)\,,\]
```
Comment (brunjlar) [inline shelley/design-spec/delegation_design_spec.tex]: I think we should mention how \eta is actually calculated (provided it is not simply the number of slots per epoch).

File: shelley/design-spec/delegation_design_spec.tex [hunk 14]
```diff
@@ -2677,38 +2716,28 @@ \subsubsection{Monetary Expansion}
...
   valuable.
 \end{itemize}
 
 Note that the fees that have been collected during the bootstrap era,
 where all nodes have been provided by IOHK, Emurgo, and the Cardano
 Foundation, have not been paid out. Those fees have reduced the amount
-of ADA currently in existence, but they did not change
+of ada currently in existence, but they did not change
 \(T_\infty\). Effectively, those fees will be distributed amongst
 future stakeholders via monetary expansion.
```
Comment (brunjlar) [inline shelley/design-spec/delegation_design_spec.tex]: Should we give a precise definition of what "ada in existence" means? Is at the sum over all outputs of the current UTxOs?

File: shelley/design-spec/delegation_design_spec.tex [hunk 15]
```diff
@@ -2718,32 +2747,108 @@ \subsection{Reward Splitting}
...
+  \[
+  p = \frac{n}{\max(N,1)}
+  \]
+  of the number \(n\) blocks it successfully added to the chain and the number
+  \(N\) of slots it was elected as a leader, and
+\item[\(r_e\)] is a factor that captures the relation between the relative stake
+  \(\sigma\) of the pool, the number \(N\) of slots it is elected as a leader,
+  and the total number \(\overline{N}\) of blocks that were added to the chain during
+  the epoch. To be precise, we have
+  \[r_e = \frac{N}{\sigma \overline{N}}\,.\]
```
Comment (brunjlar) [inline shelley/design-spec/delegation_design_spec.tex]: Strictly speaking, this fails if \overline{N} is zero. But I guess if we ever encounter an epoch that doesn't produce a single block, we have bigger problems anyway?

File: shelley/design-spec/delegation_design_spec.tex [hunk 16]
```diff
@@ -2965,31 +3025,75 @@ \subsubsection{Non-Myopic Pool Stake}
...
+\subsubsection{Apparent Performance of New Pools}
+\label{apparent-performance-of-new-pools}
+
+When a new pool is created, there will be no data to determine its apparent
+performance, which is needed to calculate its desirability and non-myopic
+rewards. In order to include a new pool in the ranked display of pools, a wallet
+will have to pick a default value for the apparent performance.
+
+In order to neither advantage nor disadvantage new pools, we recommend
+initialising the moving average by looking at the top \(k\) stake pools (i.e.,
```
Comment (brunjlar) [inline shelley/design-spec/delegation_design_spec.tex]: How do we initialize the apparent performance for the very first pools?

File: shelley/design-spec/delegation_design_spec.tex [hunk 17]
```diff
@@ -3498,7 +3603,7 @@ \subsubsection{\texorpdfstring{\(\rho\)}{\textbackslash{}rho}}
     \approx
     0.0019.
 \] This would correspond to reducing the remaining amount of available
-ADA by \({1.0019}^{73}-1\approx 0.144=14.4\%\) per year (which sounds
+ada by \({1.0019}^{73}-1\approx 0.144=14.4\%\) per year (which sounds
 awfully high\ldots).
 
 \subsubsection{\texorpdfstring{\(\tau\)}{\textbackslash{}tau}}
```
Comment (brunjlar) [inline shelley/design-spec/delegation_design_spec.tex]: The section about parameter \gamma should be removed, because \gamma is no longer used.

File: shelley/design-spec/delegation_design_spec.tex (hunk 14 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: Maybe replace "existence" by "circulation"? We don't have a T pot and rewards pot yet, currently we only have the ada in circulation, right?

File: shelley/design-spec/delegation_design_spec.tex (hunk 15 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: Yes, I guess :)

But doing the `max(1, N)` thing can't hurt.

File: shelley/design-spec/delegation_design_spec.tex (hunk 9 above)
Comment (dnadales) [reply shelley/design-spec/delegation_design_spec.tex]: Well, I like things to be as explicit as possible, but some more "colorful" prose is often necessary. It was just for my understanding anyway :)

File: shelley/design-spec/delegation_design_spec.tex (hunk 10 above)
Comment (dnadales) [reply shelley/design-spec/delegation_design_spec.tex]: Oh, I see. Thanks.

File: shelley/design-spec/delegation_design_spec.tex (hunk 13 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: Right, we should be explicit here.

I'll clarify that for Classic, it's the number of slots, for Praos the number of slots times the active slot coefficient.

File: shelley/design-spec/delegation_design_spec.tex (hunk 15 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: We don't want to add insult to injury and divide by zero if the world has ended.

File: shelley/design-spec/delegation_design_spec.tex (hunk 16 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: Right, we need to bootstrap 🤔 

I think it's fine if we just set the performance of the pools that join during the first couple of epochs (maybe while the decentralisation switch is very close to centralised, where it's harder to measure the performance) to `1`.

File: shelley/design-spec/delegation_design_spec.tex (hunk 17 above)
Comment (kantp) [reply shelley/design-spec/delegation_design_spec.tex]: Good catch, thanks!

File: shelley/design-spec/delegation_design_spec.tex (hunk 13 above)
Comment (brunjlar) [reply shelley/design-spec/delegation_design_spec.tex]: Sounds good!

File: shelley/design-spec/delegation_design_spec.tex (hunk 17 above)
Comment (brunjlar) [reply shelley/design-spec/delegation_design_spec.tex]: Very welcome!
